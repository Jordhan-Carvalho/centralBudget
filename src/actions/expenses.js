import db from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  });

    
  export const startAddExpense = (expenseData = {}) => {
    // only works bc redux middleware (redux thunk)
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0,
      } = expenseData;
      const expense = {description, note, amount, createdAt};
      //save the data to firebase (aynsc call) (used the return to testing porpouse)
        const ref = await db.ref(`users/${uid}/expenses`).push(expense);
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    };
  };
  
  // REMOVE_EXPENSE
  export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
  });

  export const startRemoveExpense = (expense) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
        const id = expense.id;
  try{
  await db.ref(`users/${uid}/expenses/${id}`).remove();
        dispatch(removeExpense(expense));
      } catch (error) {
        alert(error);
      }
      };
      };
  
  // EDIT_EXPENSE
  export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
  });

  export const startEditExpense = (id, updates) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      try{
        await db.ref(`users/${uid}/expenses/${id}`).update(updates);
        dispatch(editExpense(id, updates));
      } catch (error) {
        console.log(error);
      }
    };
  };

  // SET_EXPENSES
  export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
  });

  // export const startSetExpenses
  export const startSetExpenses = () => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
  // 1. fetch all expense data once
  //2. Parse the data to array
  const expenses = []; 
  await db.ref(`users/${uid}/expenses`).once('value', (snapshot) => {
        snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
        });
        // 3. dispatch SET_EXPENSES
        dispatch(setExpenses(expenses));
      };
      };