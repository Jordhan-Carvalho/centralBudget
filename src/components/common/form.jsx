import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class Form extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescChange = e => {
    const description = e.target.value;
    this.setState({ description });
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState({ note });
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({ amount });
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState({ createdAt });
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState({ calendarFocused: focused });
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      const error = "Please provide descp and amount";
      this.setState({ error });
    } else {
      this.setState({ error: "" });
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <label htmlFor="title">Title</label>
            <input
              id="title"
              className="form-control"
              type="text"
              placeholder="Description"
              value={this.state.description}
              onChange={this.onDescChange}
              autoFocus
            />
          </div>
          <div className="form-group">
            <label htmlFor="amount">Amount</label>
            <input
              id="amount"
              className="form-control"
              type="text"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
            />
          </div>
          <div className="form-group">
            <SingleDatePicker
              date={this.state.createdAt}
              onDateChange={this.onDateChange}
              focused={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
          <div className="form-group">
            <label htmlFor="notes">Notes</label>
            <textarea
              id="notes"
              className="form-control"
              placeholder="Add a note for your expense"
              value={this.state.note}
              onChange={this.onNoteChange}
            />
          </div>
          <button className="btn btn-dark">Add Transaction</button>
        </form>
      </div>
    );
  }
}

export default Form;
