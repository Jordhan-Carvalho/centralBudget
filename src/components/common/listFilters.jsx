import React, { Component } from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from "../../actions/filters";

export class ListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState({ calendarFocused });
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    e.target.value === "date"
      ? this.props.sortByDate()
      : this.props.sortByAmount();
  };

  render() {
    return (
      <div className="form-row align-items-center">
        <div className="col-auto">
          <input
            className="form-control"
            placeholder="Search"
            type="text"
            value={this.props.filters.text}
            onChange={this.onTextChange}
          />
        </div>
        <div className="col-auto">
          <select
            className="form-control"
            value={this.props.filters.sortBy}
            onChange={this.onSortChange}
          >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
        </div>
        <div className="col-auto">
          <DateRangePicker
            startDate={this.props.filters.startDate}
            endDate={this.props.filters.endDate}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            showClearDates={true}
          />
        </div>
      </div>
    );
  }
}

// redux call
const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ListFilters);
