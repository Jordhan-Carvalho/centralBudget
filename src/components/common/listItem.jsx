import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ListItem = props => {
  return (
    <table className="table">
      <thead className="thead-dark">
        <tr>
          <th scope="col">Title</th>
          <th scope="col">Amount</th>
          <th scope="col">Note</th>
          <th scope="col">Date</th>
          <th scope="col">Edit</th>
        </tr>
      </thead>
      <tbody>
        {props.expenses.map(obj => (
          <tr key={obj.id}>
            <td>{obj.description}</td>

            <td>{numeral(obj.amount / 100).format("$0,0.00")}</td>
            <td>{obj.note}</td>
            <td>{moment(obj.createdAt).format("MMMM Do, YYYY")}</td>

            <td>
              <Link to={`/edit/${props.type}/${obj.id}`}>Edit</Link>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default ListItem;
