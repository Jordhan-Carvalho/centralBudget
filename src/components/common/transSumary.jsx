import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import selectVisible from "../../selectors/visibleTrans";
import getTotal from "../../selectors/transaction-total";

export const TransSumary = props => {
  return (
    <React.Fragment>
      {props.type === "expense" ? (
        <p>
          Viewing {props.expenses.length}{" "}
          {props.expenses.length === 1 ? "expense" : "expenses"}, totalling{" "}
          {numeral(getTotal(props.expenses) / 100).format("$0,0.00")}
        </p>
      ) : (
        <p>
          Viewing {props.incomes.length}{" "}
          {props.incomes.length === 1 ? "income" : "incomes"}, totalling{" "}
          {numeral(getTotal(props.incomes) / 100).format("$0,0.00")}
        </p>
      )}
    </React.Fragment>
  );
};

//redux
const mapStateToProps = state => {
  return {
    expenses: selectVisible(state.expenses, state.filters),
    incomes: selectVisible(state.incomes, state.filters)
  };
};

export default connect(mapStateToProps)(TransSumary);
