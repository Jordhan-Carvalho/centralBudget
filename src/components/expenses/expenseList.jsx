import React from "react";
import { connect } from "react-redux";
import ListItem from "../common/listItem";
import selectVisible from "../../selectors/visibleTrans";

export const ExpenseList = props => {
  return (
    <div>
      {props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        <ListItem type={"expense"} expenses={props.expenses} />
      )}
    </div>
  );
};

// redux call
const mapStateToProps = state => {
  return {
    expenses: selectVisible(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
