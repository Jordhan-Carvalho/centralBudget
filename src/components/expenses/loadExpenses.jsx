import React, { Component } from "react";
import { Link } from "react-router-dom";
import ExpenseList from "./expenseList";
import ListFilters from "../common/listFilters";
import TransSumary from "../common/transSumary";

class LoadExpenses extends Component {
  state = {};
  render() {
    return (
      <React.Fragment>
        <h1>Expense List</h1>
        <hr />
        <ListFilters />
        <TransSumary type={"expense"} />
        <hr />
        <ExpenseList />
        <hr />
        <Link to="/create">
          <button className="btn btn-dark">Add Expense</button>
        </Link>
      </React.Fragment>
    );
  }
}

export default LoadExpenses;
