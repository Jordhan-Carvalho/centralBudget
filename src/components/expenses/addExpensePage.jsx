import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startAddExpense } from "../../actions/expenses";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/expenses");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Expense</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
