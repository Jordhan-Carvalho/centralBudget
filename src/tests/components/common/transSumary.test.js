import React from 'react';
import { shallow } from 'enzyme';
import { TransSumary } from '../../../components/common/transSumary';
import expenses from '../../fixtures/expenses';


// Testing with expenses
test('should render TransSumary with expenses', () => {
    const wrapper = shallow(<TransSumary expenses={expenses} type={'expense'}/>);
    expect(wrapper).toMatchSnapshot();
});


// Testing w/o expenses
test('should render TransSumary in caso of no expenses', () => {
    const wrapper = shallow(<TransSumary expenses={[]} type={'expense'}/>);
    expect(wrapper).toMatchSnapshot();
});
