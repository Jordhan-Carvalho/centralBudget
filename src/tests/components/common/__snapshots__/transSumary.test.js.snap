// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TransSumary in caso of no expenses 1`] = `
<Fragment>
  <p>
    There's 
    0
     
    expenses
    , totalling 
    $0.00
  </p>
</Fragment>
`;

exports[`should render TransSumary with expenses 1`] = `
<Fragment>
  <p>
    There's 
    3
     
    expenses
    , totalling 
    $1,141.95
  </p>
</Fragment>
`;
